Add new map control `querybydrawpolygon`

volterra79 opened the pull request:
Closes: #340 

To use a new Query by Draw Polygon and query layers by drawing a Polygon,  you need to follow these steps:

1. Add new map control on g3w-admin:
-  click on tools icon (1)
-  click on  Administration Django .. option

![Screenshot from 2023-02-13 17-05-40](https://user-images.githubusercontent.com/1051694/218509983-3c339f27-132c-4b0a-ab2f-7ea37d4f2a9e.png)

2. Click on Map controls:

![Screenshot from 2023-02-13 17-09-19](https://user-images.githubusercontent.com/1051694/218510781-84f62d74-38b4-47d7-b17b-7f3f0f92d426.png)

3.  Click on ADD MAP CONTROL

![Screenshot from 2023-02-13 17-10-45](https://user-images.githubusercontent.com/1051694/218511132-c60882bf-ef71-43b6-b6c0-3a38d32d99bb.png)

4. Write **querybydrawpolygon** as name and save it:

![Screenshot from 2023-02-13 17-12-19](https://user-images.githubusercontent.com/1051694/218511550-7c5b5735-196f-4f3a-9fdc-625cb9ad9d69.png)

5. On g3w-admin select a Group (group of Projects) and add querybydrawpolygon on **Mapcontrols** section 

![Screenshot from 2023-02-13 17-16-10](https://user-images.githubusercontent.com/1051694/218512547-f9d9843b-a47b-4a74-ad24-239b919023e9.png)

6. Open a project of above Group. You'll see new ma control button:

![Screenshot from 2023-02-13 17-02-34](https://user-images.githubusercontent.com/1051694/218509490-0b3a0c97-986c-42c0-ae49-f4a2667a27e6.png)

7. Click on control e draw a Polygon:

![Screenshot from 2023-02-13 17-19-46](https://user-images.githubusercontent.com/1051694/218513268-39ee63b2-9d4d-49bd-a4c8-a42dbc7404a5.png)

volterra79 committed:
Add querybydrawpolygon map control

volterra79 committed:
In case of no polygon feature comeing from layer but is draw (in case of mapcontrol querybydrawpolygon) empty message is set and show only draw icon

volterra79 committed:
In case of draw polygon feature and not coming from layer, polygon Layer is set to null

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Add @since jsdoc

volterra79 committed:
Fix it translation

volterra79 committed:
Create base query polygon control for common methods and attributes between querybypolygoncontrol.js and querybydrawpolygoncontrol.js

volterra79 committed:
Rename listenLayersVisibleChange to listenLayersChange for general purpose

volterra79 committed:
change for better readability

volterra79 committed:
Use basequerypolygon common module

volterra79 committed:
Add layerFilterObject to request

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

Raruto committed:
update translations

Raruto committed:
update jsdoc

Raruto committed:
code cleanup + jsdoc

Raruto committed:
remove ts-check

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

# Conflicts:
#	src/app/g3w-ol/controls/interactioncontrol.js
#	src/app/g3w-ol/controls/querybypolygoncontrol.js
#	src/app/gui/map/mapservice.js
#	src/services/data-query.js

volterra79 committed:
Add SELECTORALL filter value in case  excludeSelected is not boolean. Is used fron queryresultsservice.js

volterra79 committed:
Add querybydrawpolygon map control case

volterra79 committed:
reset state.query to null instead to empty object

volterra79 committed:
♻️ Refactor querybypolygoncontrol.js using basequerypolygoncontrol.js has base control polygon to has common code with querybydrawpolygoncontrol.js

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
♻️ Refactor querybypolygoncontrol.js querybydrawpolygoncontrol.js querybboxcontrol.js. Move all logic inside each control instead to have it on mapservice.js

volterra79 committed:
♻️ Remove change toggle property when cal setEnable. Separate two handle properties

volterra79 committed:
✨ Add clear method to clear dirty property of map control when is toggle set false

volterra79 committed:
🎨 better management of toggle method to avoid wasting time running code

volterra79 committed:
🔥 remove layer parameter in favor of layerName

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,189 @@
+const {
+  Geometry : {
+    getAllPolygonGeometryTypes,
+    isPolygonGeometryType
```
The `isPolygonGeometryType` function appears to be unused within this file: `/src/app/g3w-ol/controls/basequerypolygoncontrol.js`

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,189 @@
+    onSelectlayer,
+    enabled=true,
+    interactionClass
+  } = options;
```
Check if rewritten like so could be equivalent:

```suggestion
const BaseQueryPolygonControl = function(options = {
    spatialMethod = SPATIALMETHODS[0],
    name,
    tipLabel,
    label,
    onSelectlayer,
    enabled = true,
    interactionClass
  } = {}) {

```

volterra79 replied:
@Raruto that's wrong because:

```js
const _options = {
  offline: false,
  name,
  tipLabel,
  label,
  // update selected layer
  onSelectlayer,
  clickmap: true, // set ClickMap
  interactionClass,
  spatialMethod,
  toggledTool:{
    type: 'spatialMethod',
    how: 'toggled' // or hover
  },
  onhover: true
};
```

and for example `tipLabel` is not defined.

I rewrite it

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,189 @@
+  this.selectedLayer = layer;
+};
+
+proto.change = function(layers=[]){
+  this.layers = layers;
+  const visible = this.checkVisibile(layers);
+  this.setVisible(visible);
+  this.setEnable(false);
+  this.listenLayersVisibilityChange();
+};
```
check if the `change()` method overrides/implements another one from a parent class (e.g. `InteractionControl`) or is native to `BaseQueryPolygonControl`

Raruto committed:
re-order function declarations + make use of `@virtual` tag for declaring abstract functions

volterra79 committed:
♻️ Handle toggle and enable state of control after listen watch selected or visible state of layer.
Refactro and move common methods to interactioncontrol.js

Raruto committed:
remove nested destructuring object declaration (or whatever it else is named..)

Raruto committed:
explicit `options` destructuring

Raruto committed:
code format

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,148 @@
+   */
+  this.selectedLayer = null;
+
+  const _options = {
+    offline: false,
+    name,
+    tipLabel,
+    label,
+    // update selected layer
+    onSelectlayer,
+    clickmap: true, // set ClickMap
+    interactionClass,
+    spatialMethod,
+    toggledTool:{
+      type: 'spatialMethod',
+      how: 'toggled' // or hover
+    },
+    onhover: true
+  };
+
+  options = merge(options, _options);
+
+  options.geometryTypes = VALIDGEOMETRIES;
```
A little bit weird:

1. declare a default `_options` object with dynamic properties coming from the `option` object
2. extend the dynamic `option` object with the properties of the aforementioned default object (`_options`)
3. change again (in other places within constructor) the properties of the dynamic `options` object (ref: `options.geometryTypes` and `options.visible`)

Not so easy to figure out what could be the mandatory options..

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,148 @@
+  // starting disabled
+  this.setEnable(enabled);
+
+  this._handleExternalLayers();
```
The `_handleExternalLayers` function seems to be undefined within this file: `src/app/g3w-ol/controls/basequerypolygoncontrol.js`

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,148 @@
+   */
+  this.externalLayers = [];
+
+  /**
+   * @type {unknown[]}
+   *
+   * @since 3.8.0
+   */
+  this.unwatches = [];
```
The `unwatches` variable appears to be unusued within this file: `src/app/g3w-ol/controls/basequerypolygoncontrol.js`

volterra79 replied:
@Raruto remove

volterra79 committed:
Related to https://github.com/g3w-suite/g3w-client/pull/341#discussion_r1129149514

volterra79 committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

 Conflicts:
	src/app/g3w-ol/controls/basequerypolygoncontrol.js

volterra79 committed:
Related to https://github.com/g3w-suite/g3w-client/pull/341#discussion_r1129149514

volterra79 committed:
Revert to fix https://github.com/g3w-suite/g3w-client/pull/341#discussion_r1129070066
To improve it

Raruto committed:
refactor `Control::setEnable(enabled)`

a little step before removing jQuery selector

Raruto committed:
vertical spacing

Raruto committed:
default parameter

Raruto committed:
typedef

Raruto committed:
re-sort virtual functions

move them the top in order to avoid forgetting to implement within subclasses

Raruto committed:
add `@listens` tag + shorten variable name

volterra79 committed:
Fix removed double call of this.toggle() when map control is created

Raruto committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

Raruto committed:
add `@param`

Raruto committed:
improve documentation for `_handleExternalLayers()`

Raruto committed:
jsdoc

Raruto commented on src/app/g3w-ol/controls/querybboxcontrol.js:
```diff
@@ -1,100 +1,249 @@
 
-const QueryBBoxControl = function(options = {}){
-  const {spatialMethod=SPATIALMETHODS[0]} = options;
+// get all filtrable layers in toc no based on selection or visibility
```
Rewrite this comment better, not clear what it does (eg. `no based on selection or visibility` ?)

Raruto commented on src/app/g3w-ol/controls/querybboxcontrol.js:
```diff
@@ -1,100 +1,249 @@
-  const {spatialMethod=SPATIALMETHODS[0]} = options;
+// get all filtrable layers in toc no based on selection or visibility
+const layersFilterObject = {
+  SELECTEDORALL: true, // selected or all
```
- if possible, I think it could be better write this in a snake case (eg: `SELECTED_OR_ALL: true,`)
- for `true / false` values it is not so easy to understand what we are referring to (e.g. true --> `selected` and false --> `all` ?)
- alternatively, it could be feasible to use two different object properties or literal values (`'selected'` | `'all'`) instead of boolean values?

Raruto committed:
remove temp variable

volterra79 committed:
Handle toggled property to true when create a map control

Raruto committed:
simplify function `QueryBBoxControl::checkEnabled()`

Raruto committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

Raruto committed:
code indentation + add `@listens` tag

Raruto committed:
reduce nesting level

Raruto committed:
reduce nesting level

Raruto committed:
jsdoc style

volterra79 committed:
Move setSelectedLayer to interactioncontrol.js

volterra79 committed:
Add comment

Raruto commented on src/app/g3w-ol/controls/querybydrawpolygoncontrol.js:
```diff
@@ -0,0 +1,228 @@
+          this.setSelectedLayer(null);
+          this.setEnable(this.isThereVisibleLayers());
+          this.toggle(this.isToggled() && this.getEnable());
+        }
```
written like so could be equivalent? (I'm specifically referring to the last line `this.toggle(...)`)

```suggestion
      selected => {
        this.setSelectedLayer(true === selected ? layer : null);
        this.setEnable(true === selected ? layer.visible : this.isThereVisibleLayers());
        this.toggle(this.isToggled() && this.getEnable());
      })
```

Raruto committed:
reduce some nesting level

Raruto committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Misspelling method name

volterra79 committed:
Misspelling method name

volterra79 committed:
Move externalLayers to interactioncontrol.js

Raruto committed:
clean up requires

Raruto committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Move query handle click from mapservice.js to querycontrol.js

Raruto committed:
clean up `QueryByPolygonControl`

Raruto committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Fix setVisible after call InteractionControl

Raruto committed:
refactor function `QueryControl::setMap(map)`

Raruto committed:
move `if (map)` conditional

Raruto committed:
add alias for `queryResultsService`

volterra79 committed:
🔥 Remove unused map variable

volterra79 committed:
Fix misspelling eventKey listener property instead linstener

volterra79 committed:
Rewrite code for querycontrol.js to dispatch picked event after click on map. Need to create picked event by control because some plugins use this event to overwrite the behavior of thi event handler

Raruto committed:
missing import

Raruto commented on src/app/g3w-ol/controls/basequerypolygoncontrol.js:
```diff
@@ -0,0 +1,114 @@
+    interactionClass
+  } = options;
+
+  this.layers = options.layers || [];
+
+  this.listenLayersVisibilityChange();
+
+  /**
+   * @type {boolean}
+   */
+  options.visible = this.checkVisibile(this.layers);
+
+  const _options = {
+    offline: false,
+    name,
+    tipLabel,
+    label,
+    // update selected layer
+    onSelectlayer,
+    clickmap: true, // set ClickMap
+    interactionClass,
+    spatialMethod,
+    toggledTool:{
+      type: 'spatialMethod',
+      how: 'toggled' // or hover
+    },
+    onhover: true
+  };
+
+  options = merge(options, _options);
+
+  options.geometryTypes = VALIDGEOMETRIES;
+
+  InteractionControl.call(this, options);
+
+  // starting disabled
+  this.setEnable(enabled);
+
+  this._handleExternalLayers();
+};
```
Like so could be equivalent?

```suggestion
/**
 * @typedef BaseQueryPolygonControlOptions
 * 
 * @property {boolean}   offline
 * @property {string}    name
 * @property {string}    tipLabel
 * @property {string}    label
 * @property {() => {}}  onSelectlayer - update selected layer
 * @property {boolean}   clickmap
 * @property {prototype} interactionClass
 * @property {boolean}   onhover
 * @property {unknown[]} layers
 * @property {boolean}   visible
 * @property {string[]}  geometryTypes
 * @property { 'intersects' | 'within' }                    spatialMethod
 * @property {{ type: string, how: ('toggled' | 'hover')} } toggledTool
 */

/**
 * @param {BaseQueryPolygonControlOptions} options
 */
const BaseQueryPolygonControl = function(options = {}) {

  const default_options = {
    offline: false,
    enabled: true,
    clickmap: true,
    spatialMethod: SPATIALMETHODS[0],
    toggledTool: {
      type: 'spatialMethod',
      how: 'toggled'
    },
    onhover: true
  };

  this.layers = options.layers || [];

  this.listenLayersVisibilityChange();

  options.visible       = this.checkVisibile(this.layers);
  options.geometryTypes = VALIDGEOMETRIES;

  options = merge(default_options, options);

  InteractionControl.call(this, options);

  // starting disabled
  this.setEnable(enabled);

  this._handleExternalLayers();
};
```

Raruto commented on src/app/g3w-ol/controls/interactioncontrol.js:
```diff
@@ -252,11 +366,72 @@ proto.setLayers = function(layers=[]){
+ * @listens CatalogService~addExternalLayer
+ * @listens CatalogService~removeExternalLayer
+ *
+ * @since 3.8.0
  */
-proto.change = function(layers=[]){
-  //to owerwite to each control
+proto._handleExternalLayers = function() {
+  const CatalogService = GUI.getService('catalog');
+
+  // 0. store unwatches of extenal layers (selected or visible)
+  const unWatches = {};
+
+  // 1. update list `this.externalLayers`
+  // 2. update list `unWatches` of layer un-watchers (based on unique name of layer)
+  // 3. call `this.handleAddExternalLayer`
+  CatalogService.onafter('addExternalLayer', ({layer, type}) => {
+    if ('vector' === type) {
+      this.externalLayers.push(layer);
+      unWatches[layer.name] = [];
+      this.handleAddExternalLayer(layer, unWatches);
+    }
+  });
+
+  // 4. clean up any previously attached event listener
+  CatalogService.onafter('removeExternalLayer', ({name, type}) => {
+    if ('vector' === type) {
+      this.externalLayers = this.externalLayers.filter(layer => {
+        if (name === layer.name) {
+          this.handleRemoveExternalLayer(layer);
+          (layer === this.selectedLayer) && this.setSelectedLayer(null);
+        }
+        return name !== layer.name;
+      });
+      unWatches[name].forEach(unWatch => unWatch());
+      delete unWatches[name];
+    }
+  });
+
 };
```
This function seems to be in use only within the `BaseQueryPolygonControl` subclass

Ref: https://github.com/g3w-suite/g3w-client/pull/341#discussion_r1129147291

volterra79 committed:
🔥 Move _handleExternalLayers to interactioncontrol.js. Set externalLayers property to single point to avoid to have same layer stored to multi control

volterra79 committed:
Move base properties of basequerypolygoncontrol.js to interactioncontrol.js.

volterra79 committed:
Change some rule to enable/disable control listen selected and visible layers properties

volterra79 committed:
Move cataloglayerselected event handler controls from mapservice.js to map-controls.js store

volterra79 committed:
♻️ Move onSelectlayer from options method to control method
Rename handleAddExternalLayer to onAddExternalLayer, handleRemoveExternalLayer to onRemoveExternalLayer

volterra79 committed:
♻️ Scale control use common scales of print.

volterra79 committed:
🐛 Fix eventKey for bbox end event to easy overwrite original handler of the event

volterra79 committed:
🐛 Fix eventKey for drawned end event to easy overwrite original handler of the event

volterra79 committed:
🐛 Pass not register eventType

volterra79 committed:
Add excludeExternal param to have more control to add or not external layers from results

volterra79 committed:
Add tochighlightable property to external layer to manage tochighlightable class for example querybboxcontrol.js

volterra79 committed:
create getExternalLayers getSelectedLayer setSelectedLayer to map-controls.js to have a common point and not repeat vale for each map control

volterra79 committed:
SetSelected layer directly to map-controls.js on addExternalLayer base on selected property value.
Changed unWatches to onAddExternalLayer array related to layer and not all object.
Align all controls to use watchLayer to register layer change property

volterra79 committed:
Handle eventKey for overwrite default behavior on picked event

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Rename SELECTEDORALL in SELECTED_OR_ALL https://github.com/g3w-suite/g3w-client/pull/341/files/bc8ed246e6019ccd12e4864df721bb5cb74231ec#r1129397817

volterra79 committed:
Add utility methods

volterra79 committed:
Change excludeExternal to positive attribute name addExternal

volterra79 committed:
add external object property fro query:polygon DataService
Remove this.setSelectedLayer to querybypolygoncontrol.js

volterra79 committed:
Move out setSelectedLayer from setters

volterra79 committed:
Remove handle of cataloglayerselected from interactioncontrol.js

volterra79 committed:
Use external parameter for query:polygon data service

Raruto committed:
remove alias

let's try to avoid introducing new code unless absolutely necessary :pray:

Raruto committed:
code intendation

Raruto committed:
jsdoc

Raruto committed:
jsdoc

Raruto committed:
move virtual methods to top

Raruto committed:
move new functions to bottom

Raruto committed:
code clean up

Raruto committed:
mispelling

Raruto committed:
move new functions to bottom

Raruto committed:
remove temporary variable: `eventKey`

Raruto committed:
vertical spacing

Raruto committed:
split conditional function: `CatalogTristratTree::isHighLight()`

Raruto committed:
reduce vertical spacing

Raruto committed:
add `@TODO` and `@since` tags

Raruto committed:
remove parenthesis on computed properties

- CatalogTristateTree::_isHighLightProjectLayer
- CatalogTristateTree::_isHighLightExternalLayer

volterra79 committed:
Query Result layer feature always visible after query. Remove click action on info result to show result feature

Raruto committed:
remove unusued variable: `data = []`

Raruto committed:
code de-duplicatation

Raruto committed:
spacing

volterra79 committed:
Remove zoom to polygon feature after query request

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

Raruto committed:
remove unusued variable: `data = []`

volterra79 committed:
Remove setVisible when create control. Always visible.

volterra79 committed:
Merge branch 'v3.6_feature-query-by-draw-polygon-control' of https://github.com/g3w-suite/g3w-client into v3.6_feature-query-by-draw-polygon-control

volterra79 committed:
Fix layers for querybypolygoncontrol.js and visibility condition

volterra79 committed:
Move isThereVisibleLayers to interactioncontrol.js because it use by querybboxcontrol.js and querybydrawpolygoncontrol.js
and fix enable state of querybydrawpolygoncontrol.js

volterra79 committed:
Merge branch 'dev' into v3.6_feature-query-by-draw-polygon-control

Raruto commented on src/app/g3w-ol/controls/interactioncontrol.js:
```diff
@@ -227,36 +426,146 @@ proto._handleClick = function(evt) {
+  )
+};
+
+/**
+ * @returns {boolean} whether at least one of stored `this.layers` is visible
+ *
+ * @since 3.8.0
+ */
+proto._hasVisibleLayer = function() {
+  return !!((this.layers.length > 0) && this.layers.find(layer => layer.isVisible()));
+};
+
+/**
+ * @returns {boolean} whether at least one of stored `this.getExternalLayers()` is visible
+ *
+ * @since 3.8.0
+ */
+proto._hasVisibleExternalLayer = function() {
+  return !!(this.getExternalLayers().find(layer => layer !== this.layer && true === layer.visible));
+};
+
+/**
+ * @since 3.8.0
+ */
+proto.isThereVisibleLayers = function() {
+  return !!(this._hasVisibleLayer() || this._hasVisibleExternalLayer());
+};
```
- missing `@returns` type for `isThereVisibleLayers()` function
- consider renaming `isThereVisibleLayers()` into `hasVisibleLayers()` for greater consistency with other functions
- find out if we also need to rename some other functions as well (in order to have a easier to undestard conditions and avoid oversights) \*


\* right now it would be something like this: `hasVisibleLayers()` = `_hasVisibleLayer()` + `_hasVisibleExternalLayer()`

(so it is quite difficult to understand the difference between private `_hasVisibleLayer()` and the outer public `hasVisibleLayers()` function)

Raruto commented on src/app/g3w-ol/controls/querybydrawpolygoncontrol.js:
```diff
@@ -10,15 +10,12 @@ const { throttle }            = require('core/utils/utils');
-      }
-    });
+  const layers = GUI.getService('map').filterableLayersAvailable({
+    filtrable: {
```
before it was `FILTERABLE` instead of `filtrable` (ie. now we are missing "e"?)

Raruto replied:
That's right, edit in a subsequent pull request

volterra79 committed:
Rename methods as suggest https://github.com/g3w-suite/g3w-client/pull/341#discussion_r1138483159

volterra79 committed:
wrong filterableLayersAvailable. { FILTERABLE: true, SELECTED_OR_ALL: true } is already implemented by method

Raruto committed:
rename function: `InteractionControl::hasControlVisibleLayer` into `InteractionControl::hasVisibleProjectLayer`
